### ISSUE TITLE
[✨ feature] Create service deployments (and redeployments) for docker images services

### ISSUE THREAD

**Issue by Fredkiss3** [zane-ops/zane-ops#44]
## Is your feature request related to a problem? Please describe.

This feature is for deploying, and updating the docker image services.

## Describe the solution you'd like

We will need to create the deployment and start the service.
Each new deployment will need to pull the `latest` image from docker before running it if the image tag is `latest`, else it will just run the service with the specified image tag.

Each new deployment needs to be attached to a tag, that can be `latest` or something else.

Each deployment will be linked to a tag.
We also need to modify the `/create-service/docker` endpoint to extract the tag (if there is none, use `latest`) that will be used for the deployment.

Each deployment will correspond to the same docker service, and each deploy will update the service. 
Only one deployment can be up (the production deployment) at a time. 

Rolling back to a previous deployment means applying the updates of that deployment to the current running service. 

To make sure a service deployment (from docker) is live & succesful we do this : 
- launch the service
- collect events (wait for max 10 seconds)
   - wait that a container is up
- check that no container died during that period

If the user gives a healthheck in the request body however, we will add scheduled job and use that instead.

To get the status of a deployment, we need to also add error callbacks to the `deploy_docker_service` task so that we make it fail in that case : https://github.com/zane-ops/zane-ops/blob/aeb94beab203fca090b37289fdb30415feb677b8/backend/zane_api/tasks.py#L12

> [!IMPORTANT]
> We also need to find a way to listen for deployment services and add callback if the service is dead, maybe with a scheduled celery job 🤔 .
> **We might need to ask the user for a healthcheck endpoint or command to run periodically in the add service endpoint**, if they don't provide one, we can just check for the number of replicas and if the container is up.
>
> When we will use git deployments, we will need to attach volumes on READ-only for preview deploys because that could cause data corruption between prod & dev, we need to reference [this comment](https://www.reddit.com/r/docker/comments/hvblmq/comment/fysk28n/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1&utm_content=share_button) when the user asks why

We also need to change the deployment status to these : `PENDING`, `HEALTHY`, `UNHEALTHY`, `CANCELLED`, `SLEEPING` (for preview deploys) , `OFFLINE` (old deployments that can't be run).

We also need to disable zero-downtime deploys for services for both docker & git services with a volume attached, and replace the update action here with `stop-first` in that case : https://github.com/zane-ops/zane-ops/blob/d64fcaf5d4b7eabd819e8e3d691a041d29202a7e/backend/zane_api/docker_operations.py#L364

A user can add or remove volumes between deploys, in that case we need to make sure to create or delete one at will.

Any changes by the user to the service (except modifying the slug) will need to redeploy the service.
To monitor deploy states, we need to get swarm tasks associated with services and add a scheduled job to monitor the state.

### TLDR :
- [x] pull image tag
- [x] disable zero-downtime deploys for services with volumes
- [x] get deployment status by monitoring with a scheduled job, only the last one will matter
- [x] support healthcheck
- [x] support only ONE production deployment
- [x] support redeploys (of previous deploys)
- [x] Support deploys with updated values (should be almost the same as the service creation endpoint) 

## Additional context

- We need to make git env variables dependent of preview or prod.
- We might need to add dynamic environment variables referencing the IDs for the PR injected into each preview, so that user could use them to to interpolate and have dynamic environment variables (ex: a different db connection string per PR), we can inspire ourselves from vercel system envs : https://vercel.com/docs/projects/environment-variables/system-environment-variables#system-environment-variables
- docker swarm guide : https://gabrieltanner.org/blog/docker-swarm/#nodes

To update a service with a new image (and tag), we do this : 
- https://stackoverflow.com/a/50974560/10322846
- Dynamic periodic tasks : https://reintech.io/blog/dynamic-task-scheduling-celery
- docker healthcheck : 
   - https://dev.to/jjoc007/mastering-docker-defining-health-checks-in-docker-compose-4l5k
   - https://stackoverflow.com/questions/48570648/docker-compose-healthcheck-mysql
- https://chat.openai.com/share/4de0b995-cbe1-4091-8a02-e8b80cfad447  
- Healthcheck : https://docs.railway.app/reference/healthchecks
- Stop scripts gracelly : https://stackoverflow.com/questions/18499497/how-to-process-sigterm-signal-gracefully
- Stop celery worker : 
  - https://stackoverflow.com/questions/8920643/cancel-an-already-executing-task-with-celery
  - https://docs.celeryq.dev/en/stable/userguide/workers.html#revoke-revoking-tasks
- during a deploy that is not the initial deploy, we need to do this ? : https://docs.render.com/deploys#health-check-protocol

---
- [x] Are you willing to make a PR for this ?
